pkg/rsnotify/pqlistener/pqlistener.go

```diff
@@ -0,0 +1,204 @@
+		defer close(l.exit)
+
+		// Create a context that can be cancelled when the server shuts down.
+		ctx, cancel := context.WithCancel(context.Background())
+		l.cancel = cancel
+		defer cancel()
+
+		for {
+			if waitErr := l.wait(ctx, items, errs, ready); err != nil {
+				errs <- waitErr
+			}
+
+			// If we've been signaled to stop, exit. Otherwise, wait a second and retry
+			if needExit(ctx) {
+				return
+			}
+		}
+	}()
+
+	// Ready is notified when we are actively listening. Wait to return until
+	// we have an active listener.
+	<-ready
+	return
+}
+
+func needExit(ctx context.Context) bool {
+	tm := time.NewTimer(time.Second)
+	defer tm.Stop()
+	select {
+	case <-ctx.Done():
+		return true
+	case <-tm.C:
+	}
+	return false
+}
+
+func (l *PqListener) wait(ctx context.Context, items chan listener.Notification, errs chan error, ready chan struct{}) (err error) {
+	// Set up the connection
+	err = l.acquire(ready)
+	if err != nil {
+		return
+	}
+
+	// Listen/Notify event loop
+	ch := l.conn.NotificationChannel()
+	for {
+		select {
+		case n := <-ch:
+			l.notify(n, l.t, errs, items)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (l *PqListener) acquire(ready chan struct{}) (err error) {
+	if l.conn != nil {
+		l.conn.Unlisten(l.name)
+		l.conn.Close()
+		l.conn = nil
+	}
+
+	l.ip = ""
+	l.conn, err = l.factory.NewListener()
+	if err != nil {
+		return fmt.Errorf("error acquiring native connection: %s", err)
+	}
+
+	// Get the connection's IP
+	l.ip, err = l.factory.IP()
+	if err != nil {
+		return
+	}
+
+	err = l.conn.Listen(l.name)
+	if err != nil {
+		return fmt.Errorf("error starting listener: %s", err)
+	}
+
+	// If there were no errors, we are now actively listening. Signal that
+	// we are ready, or log when reconnecting.
+	select {
+	case <-ready:
+		// Already closed. This means that we are reconnecting
+		log.Printf("successfully reconnected listener %s", l.name)
+	default:
+		// Close the `ready` channel to signal that `Listen()` can return.
+		close(ready)
+	}
+
+	return
+}
+
+func (l *PqListener) Debugf(msg string, args ...interface{}) {
+	if l.debugLogger != nil {
+		l.debugLogger.Debugf(msg, args...)
+	}
+}
+
+func (l *PqListener) notify(n *pq.Notification, i interface{}, errs chan error, items chan listener.Notification) {
+	// A notification was received! Unmarshal it into the correct type and send it.
+	var input listener.Notification
+	input = reflect.New(reflect.ValueOf(i).Elem().Type()).Interface().(listener.Notification)
+	payloadBytes := []byte(n.Extra)
+	err := json.Unmarshal(payloadBytes, input)
+	if err != nil {
+		errs <- fmt.Errorf("error unmarshalling JSON: %s", err)
+		return
+	}
+	if unmarshaller, ok := l.unmarshallers[input.Type()]; ok {
+		err = unmarshaller(input, payloadBytes)
+		if err != nil {
+			errs <- fmt.Errorf("error unmarshalling with custom unmarshaller: %s", err)
+			return
+		}
+		l.Debugf("Unmarshalled notification of type %d with registered unmarshaller", input.Type())
```

Comment:
Should we mention the unmarshaller type here? Might be valuable information.

Reply:
An unmarshaller is a function, so I'm not sure that would be useful: https://github.com/rstudio/platform-lib/blob/4cf2da70d4cbc461b6bd0480e347ae78a718cd89/pkg/rsnotify/listener/listener.go#L32.

That said, maybe we can eventually work on a way to name the unmarshallers for better reporting.